### PR TITLE
Making `wires` the last positional argument for several classes

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** Currently, as noticed in [this issue](https://github.com/PennyLaneAI/pennylane/issues/5521#issuecomment-2092520332), `qml.GateFabric`, `qml.AllSinglesDoubles`, and `qml.BasisRotation` do not have `wires` as the last positional argument, which is a requirement for pennylane operations. This might be confusing for users.

**Description of the Change:** In `qml.GateFabric`, we made `init_state` a `kwarg` instead of a positional argument, with an initial default state as the all-zeros state. For `qml.AllSinglesDoubles` and `qml.BasisRotation` we simply changed the argument order. Now, `wires` is the rightmost positional argument in all three cases.

**Benefits:** Now, the three Python classes above have an argument order in accordance with pennylane requirements. Additionally, now `qml.AllSinglesDoubles` also accepts a `list` as `hf_state` parameter (not necessarily a `numpy` array).

**Possible Drawbacks:** In the case of `qml.AllSinglesDoubles` and `qml.BasisRotation`, this is technically a breaking change since we changed the argument order:

```
import pennylane as qml
import numpy as np

weights = np.ones(3)
singles = [[0, 2], [1, 3]]
doubles = [[0, 1, 2, 3]]
hf_state = [1, 1, 0, 0]
wires=[0, 1, 2, 3]

# works fine
qml.AllSinglesDoubles(weights, hf_state, wires, singles=singles, doubles=doubles)

# works fine
qml.AllSinglesDoubles(weights=weights, wires=wires, hf_state=hf_state, singles=singles, doubles=doubles)

# fails since the second argument is now interpreted as hf_state
# qml.AllSinglesDoubles(weights, wires, hf_state, singles=singles, doubles=doubles)
```

**Related GitHub Issues:** #5521

[sc-61430]

### CODE REVIEW

File: pennylane/templates/layers/gate_fabric.py [hunk 1]
```diff
@@ -67,9 +67,9 @@ class GateFabric(Operation):
         weights (tensor_like): Array of weights of shape ``(D, L, 2)``\,
             where ``D`` is the number of gate fabric layers and ``L = N/2-1``
             is the number of :math:`\hat{Q}(\theta, \phi)` gates per layer with N being the total number of qubits.
-        wires (Iterable): wires that the template acts on
-        init_state (tensor_like): init_state (tensor_like): iterable of shape ``(len(wires),)``\, representing the input Hartree-Fock state
-            in the Jordan-Wigner representation.
+        wires (Iterable): wires that the template acts on.
+        init_state (tensor_like): iterable of shape ``(len(wires),)``\, representing the input Hartree-Fock state
+            in the Jordan-Wigner representation. If ``None``, a tuple of zeros is selected as initial state. Default is ``None``.
```
Comment (dwierichs) [inline pennylane/templates/layers/gate_fabric.py:72]: Just an opinion:
```suggestion
            in the Jordan-Wigner representation. If ``None`` (default), a tuple of zeros is selected as initial state.
```

File: pennylane/templates/layers/gate_fabric.py [hunk 2]
```diff
@@ -67,9 +67,9 @@ class GateFabric(Operation):
         weights (tensor_like): Array of weights of shape ``(D, L, 2)``\,
             where ``D`` is the number of gate fabric layers and ``L = N/2-1``
             is the number of :math:`\hat{Q}(\theta, \phi)` gates per layer with N being the total number of qubits.
-        wires (Iterable): wires that the template acts on
-        init_state (tensor_like): init_state (tensor_like): iterable of shape ``(len(wires),)``\, representing the input Hartree-Fock state
```
Comment (dwierichs) [inline pennylane/templates/layers/gate_fabric.py:71]: :+1:

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -65,6 +65,10 @@
 
 <h3>Breaking changes 💔</h3>
 
+* `qml.GateFabric`, `qml.AllSinglesDoubles`, and `qml.BasisRotation` now have the `wires` argument as the last positional argument,
+  which is a requirement for PennyLane operations.
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:69]: I'm not sure we have a document that fixes this as a hardcore requirement. We could also call it a convention here, not sure.

**Review by dwierichs — CHANGES_REQUESTED** [review]
Looks good to me, thanks for the quick fix @PietropaoloFrisoni ! :100:

I might have overlooked it, but is there any test where `BasisRotation` is instantiated with `unitary_matrix` as positional argument? Might be good to have that in as well, just by changing some arbitrary test. Maybe the `test_standard_validity` test :)

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (PietropaoloFrisoni) [reply doc/releases/changelog-dev.md]: I was also unsure. I replaced `requirement` with `convention`

**Comment by PietropaoloFrisoni** [comment]
Thanks @dwierichs. Actually, I changed the order of the arguments in the tests as well (even where it was not technically necessary) just to be consistent. in `test_standard_validity` I noticed just now that there was a `weights` variable instead of `unitary_matrix`, so I'll replace it :+1:

File: tests/templates/test_subroutines/test_basis_rotation.py [hunk 4]
```diff
@@ -24,13 +24,13 @@
...
     """Run standard tests of operation validity."""
-    weights = np.array(
+    unitary_matrix = np.array(
         [
             [-0.618452, -0.68369054 - 0.38740723j],
             [-0.78582258, 0.53807284 + 0.30489424j],
         ]
     )
-    op = qml.BasisRotation(wires=range(2), unitary_matrix=weights)
+    op = qml.BasisRotation(unitary_matrix=unitary_matrix, wires=range(2))
```
Comment (dwierichs) [inline tests/templates/test_subroutines/test_basis_rotation.py:33]: This is the suggestion I meant:
```suggestion
    op = qml.BasisRotation(unitary_matrix, wires=range(2))
```

**Review by dwierichs — APPROVED** [review]
Thanks @PietropaoloFrisoni I iterated my suggestion. It's good to go from my side either way.

File: tests/templates/test_subroutines/test_basis_rotation.py (hunk 4 above)
Comment (PietropaoloFrisoni) [reply tests/templates/test_subroutines/test_basis_rotation.py:33]: Oh, now I see what you mean. Thanks!

File: pennylane/templates/subroutines/all_singles_doubles.py [hunk 5]
```diff
@@ -141,6 +141,9 @@ def __init__(self, weights, wires, hf_state, singles=None, doubles=None, id=None
         if weights_shape != exp_shape:
             raise ValueError(f"'weights' tensor must be of shape {exp_shape}; got {weights_shape}.")
 
+        # convert hf_state to numpy array in case it is a list
+        hf_state = np.array(hf_state) if isinstance(hf_state, list) else hf_state
```
Comment (albi3ro) [inline pennylane/templates/subroutines/all_singles_doubles.py:145]:
```suggestion
        hf_state = np.array(hf_state) if isinstance(hf_state, (list, tuple)) else hf_state
```

Comment (PietropaoloFrisoni) [reply pennylane/templates/subroutines/all_singles_doubles.py:145]: BTW, it looks like all PRs are currently failing due to `catalyst`, so I'll commit the change above although the resulting PR will most probably fail

**Review by albi3ro — COMMENTED** [review]
This PR is looking good, and I'm sorry to bring this up right now, but I'd like to propose another option for `AllSinglesDoubles`.  

We coudl instead default `hf_state=None`, and then if `hf_state is None`, we simply don't include the `BasisState` op in the decomposition.

**Comment by PietropaoloFrisoni** [comment]
@albi3ro No problem, I am fine with everything the @PennyLaneAI/algorithms is happy with

**Comment by PietropaoloFrisoni** [comment]
Closing the PR since it has been decided not to fix #5521

**Pull request closed without merging**